### ISSUE TITLE
--skip-msys option does not affect packages in snapshot

### ISSUE THREAD

**Issue by mwu-tow** [commercialhaskell/stack#1405]
The MSYS2 is breaking my builds because Cabal repeats include/library paths for MSYS2 for every package in dependencies, easily triggering overlong command problem on Windows (exceeding 32k command length).

Therefore, I need to build with `--skip-msys`. When I give the flag for the first time, it recompiles all local packages because the extra-include-dir and extra-lib-dir are changed. However, the option does not affect packages installed in my snapshot database. If they were installed with MSYS2 enabled, the include/libraries path for MSYS will come into my build, Cabal will replicate them many times (once for each dependency) and build will fail.

If presence of MSYS affects the shape of compiled package (they needs to be recompiled), than I believe there should be separate snapshot databases for MSYS and non-MSYS packages. 
Without that I have to delete my snapshot database every time I want to switch the option.  Not to say about possible problems if the msys and non-msys packages get mixed. 
[even single MSYS-built library can pollute paths with MSYS2 libraries/tools that are often incompatible with GHC.]

**Comment by borsboom** [comment]
I can see how being without msys is almost a different platform, so perhaps `--skip-msys` should adjust the `platformRelDir` which would have the effect of using separate databases.  I don't know enough about the use case `--skip-msys` was designed for to know if this is sensible though.  Will using `--skip-msys` as a workaround for the overlong command-line still be necessary once GHC/Cabal versions that support response files for all this are released?

@snoyberg, opinion?

**Comment by snoyberg** [comment]
I'm OK with making such a change. The `--skip-msys` was intended for when the user already had MSYS installed, not for the use case described in this issue (which hopefully will be better addressed with GHC 7.10.3). It's also possible to simply delete the $APPDATA\stack directory to work around this.

Rereading the initial description, I'm not seeing any concrete bug that has been encountered by mixing builds with and without the `--skip-msys` flag. Have such issues come up?

**Comment by mwu-tow** [comment]
@snoyberg 
I'm not saying that "mixing" MSYS and no-MSYS packages causes any problems by itself. It is just that having at least one MSYS-built package means that the whole build will have MSYS include/lib directories visible and the build will use MSYS even if `--skip-msys` flag is given. And using MSYS is dangerous for a few reasons and should be avoided whenever possible. At least on GHC < 7.10.3 (nut sure if the loading order fixes for GHC runtime loader got merged into 7.10.3?), not to say that we're still stuck on 7.8.4.

**Comment by snoyberg** [comment]
I disagree with the assertion of MSYS being dangerous, I haven't seen any
explanation for that claim. I think Stack's behavior here is the right one
for the general case. If you have specific reasons to avoid MSYS, you're
quite able to do so. But most users do better with MSYS than without.

**Comment by mwu-tow** [comment]
Disclaimer: Please note that I'm not discussing the MSYS itself but using MSYS environment to build Haskell programs on Windows. MSYS itself does its job well. However, I think that using it to build Haskell is generally troubling and should not be used by default. I'm not suggesting changing anything here, I'm fine as long as I can disable it.

I'm writing because I don't like leaving some claims not supported by any explanations… and maybe someone struggling with the same issues will find the descriptions below helpful.

# TLDR
MSYS is usually harmless and not visible. Saying from my experience, it is harmful more often than useful.

# Explanation
[sorry if that sounds somewhat like a rant. I spent too much time on debugging these issues.]
MSYS can be useful when you:
a) are knowledgeable with both Windows and Linux quirks
Do you really undersand how pkg-config works? What rules it uses replace prefix directory? That there are various pkg-config builds, some doing "the useful thing" and some that do not? Are you sure how quotes in command arguments are handled by the tools? There are various ports of them, incompatible with each other. [Having past experiences with other variants of Linux-like environemnt on Windows can cause surprises.] Should paths be like C:\foo or /c/foo? And so on… Generally both systems have their quirks, MSYS mixes them together. The resulting environment looks for me like some kind of vertex of strange edge-cases.
b) have deep understanding of Haskell build process
You won't be able to understand what's going on or even ask the proper question if you don't know how building your package works — from stack to Setup.hs and the exact Cabal library behavior. Build failure diagnostics are usually useless, you need to get used to reading -v3 output of Cabal.
c) are lucky
Your package uses TH or its dependency uses TH? Unfortunate. Your package has "too many" transitive dependencies? Unfortunate. You want to redistribute your program, so it can be run on other machines? Unfortunate. And so on… MSYS brings a number of opportunities for things to fail.

I never had a case when using MSYS helped me to build a Haskell package [1]. I had a few cases when using MSYS broke my builds or packaging, often in subtle and hard to debug ways.
I repeat — I'm not saying that MSYS itself is at fault. It is rather GHC/Cabal/friends if anything.

Consider the following issues:
## 1) The path explosion problem as described here.
You have many dependencies and depend on C library? Presence of MSYS will break your build. Cabal sums up include/library paths from all dependencies. If every dependency has MSYS paths repeated, this will trigger failures due to overlong command length.
And it won't give you sensible diagnostics — it will complain about missing libraries even if they are present. I wasted quite much time on debugging this issue.

[Well, overlong command failure is GHC bug. Generating overlong commands is Cabal's fault. Still, from my POV, MSYS broke my build.]

## 2) MSYS libraries are incompatible with GHC
Arguably, one of the reasons to use MSYS is to be able to use its package manager to install C/C++ libraries. In some cases it works. Usually it brings more troubles than it solve.
a) Many MSYS libraries contain both static and dynamic (shared) build. `ld` will detect and use the dynamic one, GHC dynamic loader (triggered by TH usage) will pick the static one… and undebuggable mess will occur.
b) Some MSYS libraries contains pkg-config files that cause pkg-config to generate Unix-style paths. GHC and tools below can't handle them on Windows. Then you spend hours, learning about pkg-config, trying to understand how it generates its output, tracking problematic files…
c) MSYS libraries contains runtime dependencies on MSYS-MinGW that is different from GHC-MinGW. You end up with your application depending on both GCC 4.5/4.6 and 5.x -related libraries. Collecting good set of DLL's to ship with your binary becomes hard task. Everything becomes even worse when GHC-MinGW gets prepended before MSYS-MinGW, as its pthread library misses symbols needed by GHC libraries. It won't work. You'd need to prepare separate environment for building your application (actually stack does it), and separate one for executing it.

If you doubt, please try building a program that uses TH and depends on cairo.

***

It is possible that GHC 7.10.3 fixes a number of issues described here. We're still stuck on 7.8.4 so I haven't played much with it.

***

[1] Except for the GHC itself but it is a very special case and experience was awful anyway.

**Comment by snoyberg** [comment]
No plans to make any changes on this, closing.
